### PR TITLE
wait: Unify the implementation of the two backoff managers

### PR DESCRIPTION
**Pull request by smarterclayton**
**NOTE**: This pull is part of a series of changes that introduce new context-cancellation-aware Poll methods, reduce the surface area of the wait package to a smaller set of functions (if unused, marked private, if consolidating, marked deprecated), unify the underlying loop implementation with better testing, consolidate the backoff manager code into a smaller chunk, and in general address a number of outstanding issues.  See #115077, #115116, #115113, #115140, #115064, and #107826.

Both backoff managers moved to wait in the last few years, but have different implementation code. In general they are performing similar operations as the Poll and Until methods with subtly different code paths. As part of unifying the various poll loops in wait and ensuring they handle context cancellation consistently for use within the codebase (specifically Kubelet for the near term), converge the implementations into Backoff, and reduce some of the code complexity to make it easier to review, but do not change user visible behavior.

/kind cleanup
/priority important-soon

```release-note
The wait.NewExponentialBackoffManager and wait.NewJitteringBackoffManager methods have been replaced by a method on wait.Backoff. DelayWithReset accepts a clock (for testing) and an interval after which the current state of the backoff struct will be reset with its original value.  Please see the godoc of the new method and the change for examples of how to replace usage of this function.
```

```docs
```

### CODE REVIEW

**Comment by smarterclayton** [comment]
/hold until sufficient review of #107826 validates opinions on the structure of how `DelayFunc` works.

Still suitable for review by someone with backoff manager expertise, especially @zhan849.

**Comment by smarterclayton** [comment]
/hold cancel

Now that wait.Timer exists and covers the backoff manager use case (Step is discouraged, DelayFunc() is an iterator-like experience), this is suitable for review.  See #107826 for the full usage example.

File: staging/src/k8s.io/apimachinery/pkg/util/wait/wait.go [hunk 1]
```diff
@@ -231,14 +231,19 @@ func (cf ConditionFunc) WithContext() ConditionWithContextFunc {
...
 	ctx, cancel := context.WithCancel(context.Background())
-
-	go func() {
-		select {
-		case <-parentCh:
-			cancel()
-		case <-ctx.Done():
-		}
-	}()
+	select {
```
Comment (aojea) [inline staging/src/k8s.io/apimachinery/pkg/util/wait/wait.go:250]: @smarterclayton this PR was superseded by https://github.com/kubernetes/kubernetes/pull/115140, right?

**Comment by aojea** [comment]
I reviewed in https://github.com/kubernetes/kubernetes/pull/107826

File: staging/src/k8s.io/apimachinery/pkg/util/wait/wait.go (hunk 1 above)
Comment (smarterclayton) [reply staging/src/k8s.io/apimachinery/pkg/util/wait/wait.go:250]: Correct, let me address your review comments first and I'll update.

**Comment by smarterclayton** [comment]
Rolled this into #107826 after taking Antonio's suggestion to preserve the existing methods and deprecate them instead.  This is now *after* the unification and not a standalone change.

**Pull request closed without merging**